### ISSUE TITLE
[Support]: 0.12.0 upgrade "database is locked" (frigate.db on local nvme)

### ISSUE THREAD

**Issue by linrav** [blakeblackshear/frigate#6014]
### Describe the problem you are having

Cannot get 0.12.0 upgrade to get through migration. 
012_add_segment_size bizarrely fails with "database is locked"
the database is hosted on a local nvme that is bound into the container, 0.11.1 thankfully works just fine

Found only https://github.com/blakeblackshear/frigate/issues/594 and guides to move db off network shares in search. sorry if this is known.

I have already tried shutting down homeassistant while trying the upgrade, and also making a backup of db, removing all cameras from config and trying to upgrade. Neither individually nor combined resolved the "database locked" issue. Can't think of what else may be relevant

### Version

0.12.0

### Frigate config file

```yaml
database:
  path: /sqlite/db/frigate.db
mqtt:
  host: mqtt
  port: 1883
cameras:
  # camera config sortof sensitive / i can provide it if needed
  # ~15 cameras
```

### Relevant log output

```shell
s6-rc: info: service s6rc-fdholder: starting
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service s6rc-fdholder successfully started
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service log-prepare: starting
s6-rc: info: service log-prepare successfully started
s6-rc: info: service nginx-log: starting
s6-rc: info: service go2rtc-log: starting
s6-rc: info: service frigate-log: starting
s6-rc: info: service go2rtc-log successfully started
s6-rc: info: service go2rtc: starting
s6-rc: info: service nginx-log successfully started
s6-rc: info: service frigate-log successfully started
s6-rc: info: service go2rtc successfully started
s6-rc: info: service go2rtc-healthcheck: starting
s6-rc: info: service frigate: starting
s6-rc: info: service go2rtc-healthcheck successfully started
s6-rc: info: service frigate successfully started
s6-rc: info: service nginx: starting
2023-04-12 03:18:38.929895188  [INFO] Preparing go2rtc config...
s6-rc: info: service nginx successfully started
2023-04-12 03:18:38.931828182  [INFO] Starting Frigate...
s6-rc: info: service legacy-services: starting
2023-04-12 03:18:38.933311180  [INFO] Starting NGINX...
s6-rc: info: service legacy-services successfully started
2023-04-12 03:18:39.200106252  [INFO] Starting go2rtc...
2023-04-12 03:18:39.260258793  03:18:39.260 INF go2rtc version 1.2.0 linux/amd64
2023-04-12 03:18:39.260488274  03:18:39.260 INF [api] listen addr=:1984
2023-04-12 03:18:39.260628908  03:18:39.260 INF [rtsp] listen addr=:8554
2023-04-12 03:18:39.260818755  03:18:39.260 INF [srtp] listen addr=:8443
2023-04-12 03:18:39.260934563  03:18:39.260 INF [webrtc] listen addr=:8555
2023-04-12 03:18:39.490607065  2023/04/12 03:18:39 [error] 138#138: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.233, server: , request: "GET /api/stats HTTP/1.1", upstream: "http://127.0.0.1:5001/stats", host: "192.168.1.134:9595"
2023-04-12 03:18:39.490656879  192.168.1.233 - - [12/Apr/2023:03:18:39 -0700] "GET /api/stats HTTP/1.1" 502 157 "-" "HomeAssistant/2023.4.2 aiohttp/3.8.4 Python/3.10" "-"
2023-04-12 03:18:39.971623987  [2023-04-12 03:18:39] frigate.app                    INFO    : Starting Frigate (0.12.0-da3e197)
[2023-04-12 03:18:39] frigate.app                    INFO    : Starting Frigate (0.12.0-da3e197)
2023-04-12 03:18:40.040570437  [2023-04-12 03:18:40] frigate.app                    INFO    : Creating directory: /tmp/cache
2023-04-12 03:18:40.041861604  [2023-04-12 03:18:40] peewee_migrate                 INFO    : Starting migrations
[2023-04-12 03:18:40] peewee_migrate                 INFO    : Starting migrations
[2023-04-12 03:18:40] peewee_migrate                 INFO    : Migrate "012_add_segment_size"
[2023-04-12 03:18:40] peewee_migrate                 INFO    : add_column ('recordings', 'segment_size', <FloatField: Recordings.segment_size>)
2023-04-12 03:18:40.078833711  [2023-04-12 03:18:40] peewee_migrate                 INFO    : Migrate "012_add_segment_size"
2023-04-12 03:18:40.078910846  [2023-04-12 03:18:40] peewee_migrate                 INFO    : add_column ('recordings', 'segment_size', <FloatField: Recordings.segment_size>)
2023-04-12 03:18:44.490678350  2023/04/12 03:18:44 [error] 138#138: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.233, server: , request: "GET /api/stats HTTP/1.1", upstream: "http://127.0.0.1:5001/stats", host: "192.168.1.134:9595"
2023-04-12 03:18:44.490729836  192.168.1.233 - - [12/Apr/2023:03:18:44 -0700] "GET /api/stats HTTP/1.1" 502 157 "-" "HomeAssistant/2023.4.2 aiohttp/3.8.4 Python/3.10" "-"
2023-04-12 03:18:45.087451135  database is locked
2023-04-12 03:18:45.087619812  [2023-04-12 03:18:45] peewee_migrate                 ERROR   : Migration failed: 012_add_segment_size
2023-04-12 03:18:45.087622157  Traceback (most recent call last):
2023-04-12 03:18:45.087624301    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3237, in execute_sql
2023-04-12 03:18:45.087629100      cursor.execute(sql, params or ())
2023-04-12 03:18:45.087630873  sqlite3.OperationalError: database is locked
2023-04-12 03:18:45.087633338  
2023-04-12 03:18:45.087634951  During handling of the above exception, another exception occurred:
2023-04-12 03:18:45.087650350  
2023-04-12 03:18:45.087651772  Traceback (most recent call last):
2023-04-12 03:18:45.087653666    File "/usr/local/lib/python3.9/dist-packages/peewee_migrate/router.py", line 195, in run_one
2023-04-12 03:18:45.087655169      migrator.run()
2023-04-12 03:18:45.087656852    File "/usr/local/lib/python3.9/dist-packages/peewee_migrate/migrator.py", line 151, in run
2023-04-12 03:18:45.087672061      op.run()
2023-04-12 03:18:45.087673874    File "/usr/local/lib/python3.9/dist-packages/playhouse/migrate.py", line 164, in run
2023-04-12 03:18:45.087676048      self._handle_result(method(*self.args, **kwargs))
2023-04-12 03:18:45.087677851    File "/usr/local/lib/python3.9/dist-packages/playhouse/migrate.py", line 158, in _handle_result
2023-04-12 03:18:45.087700644      self._handle_result(item)
2023-04-12 03:18:45.087702899    File "/usr/local/lib/python3.9/dist-packages/playhouse/migrate.py", line 155, in _handle_result
2023-04-12 03:18:45.087704782      result.run()
2023-04-12 03:18:45.087707056    File "/usr/local/lib/python3.9/dist-packages/playhouse/migrate.py", line 164, in run
2023-04-12 03:18:45.087708790      self._handle_result(method(*self.args, **kwargs))
2023-04-12 03:18:45.087718448    File "/usr/local/lib/python3.9/dist-packages/playhouse/migrate.py", line 153, in _handle_result
2023-04-12 03:18:45.087719931      self.execute(result)
2023-04-12 03:18:45.087721614    File "/usr/local/lib/python3.9/dist-packages/playhouse/migrate.py", line 149, in execute
2023-04-12 03:18:45.087723117      self.migrator.database.execute(node)
2023-04-12 03:18:45.087724740    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3250, in execute
2023-04-12 03:18:45.087726132      return self.execute_sql(sql, params, commit=commit)
2023-04-12 03:18:45.087727805    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3244, in execute_sql
2023-04-12 03:18:45.087738976      self.commit()
2023-04-12 03:18:45.087740750    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3010, in __exit__
2023-04-12 03:18:45.087742343      reraise(new_type, new_type(exc_value, *exc_args), traceback)
2023-04-12 03:18:45.087744006    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 192, in reraise
2023-04-12 03:18:45.087745599      raise value.with_traceback(tb)
2023-04-12 03:18:45.087747262    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3237, in execute_sql
2023-04-12 03:18:45.087748645      cursor.execute(sql, params or ())
2023-04-12 03:18:45.087750157  peewee.OperationalError: database is locked
2023-04-12 03:18:46.301899205  [INFO] Service Frigate exited with code 1 (by signal 0)

```

### FFprobe output from your camera

```shell
not relevant / 0.11.1 works fine with my config
```

### Frigate stats

```json
(none, fails to start)
```

### Operating system

Debian (Docker)

### Install method

Docker CLI.

### Coral version

USB

### Network connection

Wired

### Camera make and model

unifi,wyze,eufy,reolink,others

### Any other information that may be helpful

I have already tried shutting down homeassistant while trying the upgrade, and also making a backup of db, removing all cameras from config and trying to upgrade. Neither individually nor combined resolved the "database locked" issue

**Comment by NickM-27** [comment]
What about just deleting the DB along with the -shm and -wal files?

Nothing changed with the way DBs are handled so I'd guess it was most likely just an error in the upgrade

**Comment by linrav** [comment]
The upgrade is just an add column and never even gets to the db. Db is fine
on v 0.11 still

**Comment by linrav** [comment]
This seems a bug though. There is nothing wrong with the db. I am able to
alter and open it and modify separately and v0.11.1 has no issues with it.
Why would there be locking during migration before any access to it
started? Sorry if it's an ignorant question.

 When i get home I can try making a backup and doing the migration manually
via sqlite3 cli to see if that pushes it through to confirm this is a bug
if it would be convincing

**Comment by NickM-27** [comment]
it needs to access the db in order to migrate the db to the new schema. 

Frigate 0.12 has had over 12,000 downloads at this point so if it was a general bug we would have a lot more reports / complaints about this.

The recommendation would be to make a backup and just delete it to ensure that frigate can start when creating a new DB. Then work can be done to see why the old DB is unhappy when being connected to by 0.12.

**Comment by linrav** [comment]
I'm not saying its a general bug. I am saying a fundamental change was
possibly made to how migration mutares the database unlike previous
iterations that doesn't like my setup 😂 it seems alarming that 0.12.0
migration can't access db when every version of frigate few years had no
issues.

My retention of clips and usage of related APIs for training is quite
long/heavy, so starting fresh would be pretty regrettable

 I'll try both manually upgrading the db as well as reproduction in a
duplicate environment with just deleting it to make sure it can start

**Comment by linrav** [comment]
I have not been able to reproduce what I saw yesterday this afternoon!! I am still using the same tag so not sure what changed.

Reviewing my server's hostmetrics/grafana did show significant disk-io on the nvme holding the `frigate.db` from some unrelated workloads, but those have always been running. Maybe I just got lucky with previous upgrades.

Regardless thanks for your time and support! Especially in this release Nicolas!
